crates/transpiler/src/passes/vf2/vf2_layout.rs

```diff
@@ -376,229 +368,75 @@ pub fn vf2_layout_pass(
-            target,
-        ))
+    let add_interaction = |count: &mut usize, _: &PackedInstruction, repeats: usize| {
+        *count += repeats;
+    };
+    let score = |count: &usize, err: &f64| -> Result<Option<f64>, Infallible> {
+        Ok(Some(*err * (*count as f64)))
+    };
+    let Some(avg_error_map) = avg_error_map.or_else(|| build_average_error_map(target)) else {
+        return Ok(None);
+    };
+    let Some(mut coupling_graph) = build_coupling_map(target, &avg_error_map) else {
+        return Ok(None);
+    };
+    if !strict_direction {
+        loosen_directionality(&mut coupling_graph);
     }
-}
-
-fn score_layout_internal<Ty: EdgeType>(
-    mapping: &HashMap<VirtualQubit, PhysicalQubit>,
-    error_map: &ErrorMap,
-    im_graph_data: &InteractionGraphData<Ty>,
-    strict_direction: bool,
-) -> PyResult<f64> {
-    let edge_filter_map = |a: NodeIndex, b: NodeIndex, gate_count: usize| -> Option<f64> {
-        let qubit_a = VirtualQubit(
-            im_graph_data.reverse_im_graph_node_map[a.index()]
-                .unwrap()
-                .0,
-        );
-        let qubit_b = VirtualQubit(
-            im_graph_data.reverse_im_graph_node_map[b.index()]
-                .unwrap()
-                .0,
-        );
-
-        let qargs = [mapping[&qubit_a], mapping[&qubit_b]];
-        let mut error = error_map.error_map.get(&qargs);
-        if !strict_direction && error.is_none() {
-            error = error_map.error_map.get(&[qargs[1], qargs[0]]);
+    let interactions = VirtualInteractions::from_dag(dag, add_interaction)?;
+    let start_time = Instant::now();
+    let mut times_up = false;
+    let mut trials: usize = 0;
+    let max_trials: usize = match max_trials {
+        Some(max_trials) => max_trials.try_into().unwrap_or(0),
+        None => {
+            15 + interactions
+                .graph
+                .edge_count()
+                .max(coupling_graph.edge_count())
         }
-        error.map(|error| {
-            if !error.is_nan() {
-                (1. - error).powi(gate_count as i32)
-            } else {
-                1.
-            }
-        })
     };
-
-    let bit_filter_map = |v_bit_index: NodeIndex, gate_counts: usize| -> Option<f64> {
-        let v_bit = VirtualQubit(
-            im_graph_data.reverse_im_graph_node_map[v_bit_index.index()]
-                .unwrap()
-                .0,
-        );
-        let p_bit = mapping[&v_bit];
-        let error = error_map.error_map.get(&[p_bit, p_bit]);
-
-        error.map(|error| {
-            if !error.is_nan() {
-                (1. - error).powi(gate_counts as i32)
-            } else {
-                1.
-            }
-        })
+    let time_limit = time_limit.unwrap_or(f64::INFINITY);
+    let Some((mapping, _score)) = vf2::Vf2Algorithm::new(
+        &interactions.graph,
+        &coupling_graph,
+        (score, score),
+        false,
+        vf2::Problem::Subgraph,
+        call_limit,
+    )
```

Comment:
This is probably a naive question, but I am a bit confused about how `(score, score)` propagates into the vf2 algorithm. I understand that both nodes and edges have the same scoring mechanism described by `let score = |count: &usize, err: &f64| -> Result<Option<f64>, Infallible> { Ok(Some(*err * (*count as f64))) };` which depends on the number of "needle" edges mapped to a "haystack" edge and the error on the "haystack" edge. But somehow I thought that the scoring mechanism in vf2 considers both "needle" and "haystack" nodes/edges when scoring the interaction.

Reply:
The 2-tuple is of `(node_scorer, edge_scorer)`, then for each closure, the left argument is the weight from the needle and the right argument is the weight from the haystack.  Since in `VF2Layout` our node and edge weights are the same as each other (number of 1q/2q operations for the needle, and average error rate of that qargs set for the haystack), we can use the same scoring function.